File: SAFTExtractor/Services/SAFTXMLDetails.cs
using System.Data;
using System.Data.SqlClient;
using SAFTExtractor.Models;
using SAFTExtractor.Utils;

namespace SAFTExtractor.Services
{
    /// <summary>
    /// Classe crítica que extrai dados das Stored Procedures e monta a estrutura SAFT
    /// Esta é a classe principal para extração de dados do PULSE DOMINOS
    /// </summary>
    public class SAFTXMLDetails
    {
        public SAFTXMLDetails()
        {
        }
        
        /// <summary>
        /// Obtém clientes para o SAFT XML
        /// </summary>
        public List<Customer> GetCustomers()
        {
            var customers = new List<Customer>();
            
            using (var connection = PulseConnection.CreateSQLConnection())
            {
                connection.Open();
                
                using (var command = new SqlCommand("spGetCustomersForSAFTXML", connection))
                {
                    command.CommandType = CommandType.StoredProcedure;
                    command.CommandTimeout = 0;
                    command.Parameters.AddWithValue("@LocationCode", SystemSettings.LocationCode);
                    // Parâmetro @InvoiceNumbers será adicionado quando necessário
                    
                    using (var reader = command.ExecuteReader())
                    {
                        while (reader.Read())
                        {
                            var customer = new Customer
                            {
                                CustomerID = reader["CustomerID"].ToString() ?? string.Empty,
                                AccountID = reader["AccountID"].ToString() ?? string.Empty,
                                CustomerTaxID = reader["CustomerTaxID"].ToString() ?? string.Empty,
                                CompanyName = reader["CompanyName"].ToString() ?? string.Empty,
                                Contact = reader["Contact"] != DBNull.Value ? reader["Contact"].ToString() : null,
                                Telephone = reader["Telephone"] != DBNull.Value ? reader["Telephone"].ToString() : null,
                                Fax = reader["Fax"] != DBNull.Value ? reader["Fax"].ToString() : null,
                                Email = reader["Email"] != DBNull.Value ? reader["Email"].ToString() : null,
                                Website = reader["Website"] != DBNull.Value ? reader["Website"].ToString() : null,
                                SelfBillingIndicator = reader["SelfBillingIndicator"] != DBNull.Value ? 
                                    Convert.ToInt32(reader["SelfBillingIndicator"]) : 0,
                                BuildingNumber = reader["BuildingNumber"] != DBNull.Value ? reader["BuildingNumber"].ToString() : null,
                                StreetName = reader["StreetName"] != DBNull.Value ? reader["StreetName"].ToString() : null,
                                AddressDetail = reader["AddressDetail"].ToString() ?? string.Empty,
                                City = reader["City"].ToString() ?? string.Empty,
                                PostalCode = reader["PostalCode"].ToString() ?? string.Empty,
                                Region = reader["Region"] != DBNull.Value ? reader["Region"].ToString() : null,
                                Country = reader["Country"].ToString() ?? "PT"
                                
                                // ADICIONE AQUI NOVOS CAMPOS:
                                // Exemplo: NIFRepresentante = reader["NIFRepresentante"] != DBNull.Value ? reader["NIFRepresentante"].ToString() : null
                            };
                            
                            customers.Add(customer);
                        }
                    }
                }
            }
            
            return customers;
        }
        
        /// <summary>
        /// Obtém produtos para o SAFT XML
        /// </summary>
        public List<Product> GetProducts()
        {
            var products = new List<Product>();
            
            using (var connection = PulseConnection.CreateSQLConnection())
            {
                connection.Open();
                
                using (var command = new SqlCommand("spGetProductsForSAFTXML", connection))
                {
                    command.CommandType = CommandType.StoredProcedure;
                    command.CommandTimeout = 0;
                    command.Parameters.AddWithValue("@LocationCode", SystemSettings.LocationCode);
                    // Parâmetro @ProductCodes será adicionado quando necessário
                    
                    using (var reader = command.ExecuteReader())
                    {
                        while (reader.Read())
                        {
                            var product = new Product
                            {
                                ProductType = reader["ProductType"].ToString() ?? "P",
                                ProductCode = reader["ProductCode"].ToString() ?? string.Empty,
                                ProductGroup = reader["ProductGroup"] != DBNull.Value ? reader["ProductGroup"].ToString() : null,
                                ProductDescription = reader["ProductDescription"].ToString() ?? string.Empty,
                                ProductNumberCode = reader["ProductNumberCode"].ToString() ?? string.Empty
                                
                                // ADICIONE AQUI NOVOS CAMPOS:
                                // Exemplo: CustomField = reader["CustomField"] != DBNull.Value ? reader["CustomField"].ToString() : null
                            };
                            
                            products.Add(product);
                        }
                    }
                }
            }
            
            return products;
        }
        
        /// <summary>
        /// Obtém faturas/documentos para o SAFT XML
        /// </summary>
        public List<Invoice> GetInvoices(DateTime startDate, DateTime endDate)
        {
            var invoices = new List<Invoice>();
            
            using (var connection = PulseConnection.CreateSQLConnection())
            {
                connection.Open();
                
                using (var command = new SqlCommand("spGetInvoicesForSAFTXML", connection))
                {
                    command.CommandType = CommandType.StoredProcedure;
                    command.CommandTimeout = 0;
                    command.Parameters.AddWithValue("@LocationCode", SystemSettings.LocationCode);
                    command.Parameters.AddWithValue("@StartDate", startDate);
                    command.Parameters.AddWithValue("@EndDate", endDate);
                    command.Parameters.AddWithValue("@EndDate", endDate);
                    
                    using (var reader = command.ExecuteReader())
                    {
                        while (reader.Read())
                        {
                            var invoice = new Invoice
                            {
                                InvoiceNo = reader["InvoiceNo"].ToString() ?? string.Empty,
                                ATCUD = reader["ATCUD"] != DBNull.Value ? reader["ATCUD"].ToString() : null,
                                Hash = reader["Hash"] != DBNull.Value ? reader["Hash"].ToString() : null,
                                HashControl = reader["HashControl"] != DBNull.Value ? reader["HashControl"].ToString() : null,
                                InvoiceDate = Convert.ToDateTime(reader["InvoiceDate"]),
                                SystemEntryDate = Convert.ToDateTime(reader["SystemEntryDate"]),
                                CustomerID = reader["CustomerID"].ToString() ?? string.Empty,
                                TaxPayable = Convert.ToDecimal(reader["TaxPayable"]),
                                NetTotal = Convert.ToDecimal(reader["NetTotal"]),
                                GrossTotal = Convert.ToDecimal(reader["GrossTotal"])
                                
                                // ADICIONE AQUI NOVOS CAMPOS:
                            };
                            
                            invoices.Add(invoice);
                        }
                    }
                }
                
                // Carregar linhas das faturas
                foreach (var invoice in invoices)
                {
                    invoice.Lines = GetInvoiceLines(invoice.InvoiceNo);
                }
            }
            
            return invoices;
        }
        
        /// <summary>
        /// Obtém linhas de uma fatura
        /// </summary>
        private List<InvoiceLine> GetInvoiceLines(string invoiceNo)
        {
            var lines = new List<InvoiceLine>();
            
            using (var connection = PulseConnection.CreateSQLConnection())
            {
                connection.Open();
                
                // Nota: Esta SP não existe nas SPs originais, mas é usada no código
                using (var command = new SqlCommand("spGetInvoiceLinesForSAFTXML", connection))
                {
                    command.CommandType = CommandType.StoredProcedure;
                    command.CommandTimeout = 0;
                    command.Parameters.AddWithValue("@InvoiceNo", invoiceNo);
                    
                    using (var reader = command.ExecuteReader())
                    {
                        while (reader.Read())
                        {
                            var line = new InvoiceLine
                            {
                                LineNumber = Convert.ToInt32(reader["LineNumber"]),
                                ProductCode = reader["ProductCode"].ToString() ?? string.Empty,
                                ProductDescription = reader["ProductDescription"].ToString() ?? string.Empty,
                                Quantity = Convert.ToDecimal(reader["Quantity"]),
                                UnitOfMeasure = reader["UnitOfMeasure"].ToString() ?? "UN",
                                UnitPrice = Convert.ToDecimal(reader["UnitPrice"]),
                                TaxType = reader["TaxType"].ToString() ?? "IVA",
                                TaxCountryRegion = reader["TaxCountryRegion"].ToString() ?? "PT",
                                TaxCode = reader["TaxCode"].ToString() ?? "NOR",
                                TaxPercentage = Convert.ToDecimal(reader["TaxPercentage"]),
                                TaxExemptionReason = reader["TaxExemptionReason"] != DBNull.Value ? 
                                    reader["TaxExemptionReason"].ToString() : null,
                                TaxExemptionCode = reader["TaxExemptionCode"] != DBNull.Value ? 
                                    reader["TaxExemptionCode"].ToString() : null,
                                CreditAmount = Convert.ToDecimal(reader["CreditAmount"])
                                
                                // ADICIONE AQUI NOVOS CAMPOS:
                            };
                            
                            lines.Add(line);
                        }
                    }
                }
            }
            
            return lines;
        }
        
        /// <summary>
        /// Obtém header/configurações da empresa
        /// </summary>
        public SAFTHeader GetHeader(int fiscalYear)
        {
            try
            {
                using (var connection = PulseConnection.CreateSQLConnection())
                {
                    connection.Open();
                    
                    // Usar o nome correto da SP: spGetSAFTXMLHeaderDetails
                    using (var command = new SqlCommand("spGetSAFTXMLHeaderDetails", connection))
                    {
                        command.CommandType = CommandType.StoredProcedure;
                        command.CommandTimeout = 0;
                        command.Parameters.AddWithValue("@LocationCode", SystemSettings.LocationCode);
                        
                        using (var reader = command.ExecuteReader())
                        {
                            if (reader.Read())
                            {
                                return new SAFTHeader
                                {
                                    AuditFileVersion = reader["AuditFileVersion"].ToString() ?? "1.04_01",
                                    CompanyID = reader["CompanyID"].ToString() ?? string.Empty,
                                    TaxRegistrationNumber = reader["TaxRegistrationNumber"].ToString() ?? string.Empty,
                                    TaxAccountingBasis = reader["TaxAccountingBasis"].ToString() ?? "F",
                                    CompanyName = reader["CompanyName"].ToString() ?? string.Empty,
                                    BusinessAddress = new CompanyAddress
                                    {
                                        BuildingNumber = reader["BuildingNumber"] != DBNull.Value ? reader["BuildingNumber"].ToString() : null,
                                        StreetName = reader["StreetName"] != DBNull.Value ? reader["StreetName"].ToString() : null,
                                        AddressDetail = reader["AddressDetail"].ToString() ?? string.Empty,
                                        City = reader["City"].ToString() ?? string.Empty,
                                        PostalCode = reader["PostalCode"].ToString() ?? string.Empty,
                                        Region = reader["Region"] != DBNull.Value ? reader["Region"].ToString() : null,
                                        Country = reader["Country"].ToString() ?? "PT"
                                    },
                                    FiscalYear = new DateTime(fiscalYear, 1, 1),
                                    StartDate = SAFTDate.GetFiscalYearStartDate(fiscalYear),
                                    EndDate = SAFTDate.GetFiscalYearEndDate(fiscalYear),
                                    CurrencyCode = reader["CurrencyCode"].ToString() ?? "EUR",
                                    DateCreated = DateTime.Now,
                                    TaxEntity = reader["TaxEntity"].ToString() ?? "Global",
                                    ProductCompanyTaxID = reader["ProductCompanyTaxID"].ToString() ?? string.Empty,
                                    SoftwareCertificateNumber = reader["SoftwareCertificateNumber"].ToString() ?? string.Empty,
                                    ProductID = reader["ProductID"].ToString() ?? "SAFTExtractor/PULSE DOMINOS",
                                    ProductVersion = reader["ProductVersion"].ToString() ?? "1.0"
                                };
                            }
                            else
                            {
                                // SP executou mas não retornou dados
                                throw new Exception(
                                    $"A Stored Procedure 'spGetSAFTXMLHeaderDetails' não retornou dados.\n\n" +
                                    $"LocationCode usado: '{SystemSettings.LocationCode}'\n\n" +
                                    $"Verifique se:\n" +
                                    $"1. O LocationCode '{SystemSettings.LocationCode}' existe na base de dados\n" +
                                    $"2. Existem dados de empresa configurados para este LocationCode\n" +
                                    $"3. A SP está a retornar os campos corretos"
                                );
                            }
                        }
                    }
                }
            }
            catch (SqlException sqlEx)
            {
                // Erro específico de SQL Server
                if (sqlEx.Number == 2812) // SP não encontrada
                {
                    throw new Exception(
                        $"ERRO: Stored Procedure 'spGetSAFTXMLHeaderDetails' não foi encontrada!\n\n" +
                        $"A SP precisa existir na base de dados para a aplicação funcionar.\n\n" +
                        $"Verifique a pasta 'Original SP' no repositório para obter o código SQL.\n\n" +
                        $"Detalhes técnicos: {sqlEx.Message}",
                        sqlEx
                    );
                }
                else if (sqlEx.Number == 4060) // Base de dados não existe
                {
                    throw new Exception(
                        $"ERRO: Base de dados não encontrada!\n\n" +
                        $"Connection String: {PulseConnection.GetConnectionStringSafe()}\n\n" +
                        $"Verifique o App.config e confirme que a base de dados existe.\n\n" +
                        $"Detalhes técnicos: {sqlEx.Message}",
                        sqlEx
                    );
                }
                else if (sqlEx.Number == 18456) // Login failed
                {
                    throw new Exception(
                        $"ERRO: Falha na autenticação!\n\n" +
                        $"Verifique as credenciais no App.config.\n" +
                        $"Se usar Integrated Security, confirme que tem permissões.\n\n" +
                        $"Detalhes técnicos: {sqlEx.Message}",
                        sqlEx
                    );
                }
                else
                {
                    throw new Exception(
                        $"ERRO SQL ao obter header SAFT:\n\n" +
                        $"LocationCode: '{SystemSettings.LocationCode}'\n" +
                        $"Código erro SQL: {sqlEx.Number}\n" +
                        $"Mensagem: {sqlEx.Message}\n\n" +
                        $"Verifique a configuração da base de dados no App.config.",
                        sqlEx
                    );
                }
            }
            catch (Exception ex) when (ex.Message.Contains("spGetSAFTXMLHeaderDetails"))
            {
                // Re-throw se já for uma exceção que tratamos
                throw;
            }
            catch (Exception ex)
            {
                // Erro genérico
                throw new Exception(
                    $"ERRO ao conectar à base de dados PULSE DOMINOS:\n\n" +
                    $"Connection String: {PulseConnection.GetConnectionStringSafe()}\n" +
                    $"LocationCode: '{SystemSettings.LocationCode}'\n\n" +
                    $"Verifique:\n" +
                    $"1. O SQL Server está em execução\n" +
                    $"2. O App.config tem a connection string correta\n" +
                    $"3. Tem permissões para aceder à base de dados\n\n" +
                    $"Detalhes técnicos: {ex.Message}",
                    ex
                );
            }
        }
    }
}

File: saft_extractor_python/services/saft_xml_details.py
"""
Extração de dados das Stored Procedures do PULSE DOMINOS
Esta classe é crítica - é responsável por toda a extração de dados
"""
import pyodbc
from datetime import datetime
from decimal import Decimal
from typing import List, Dict
from models.saft_models import (
    SAFTHeader, CompanyAddress, Customer, Product, 
    Invoice, InvoiceLine, DocumentTotals
)
from utils.pulse_connection import PulseConnection

class SAFTXMLDetails:
    """Extrai dados das SPs e monta estrutura SAFT"""
    
    def __init__(self):
        self.location_code = PulseConnection.get_location_code()
        self.command_timeout = PulseConnection.get_command_timeout()
    
    def get_header(self, fiscal_year: int, start_date: datetime, end_date: datetime) -> SAFTHeader:
        """
        Obtém header combinando dados de 2 SPs:
        1. spGetSAFTDetails - Retorna Setting/Value (quase todo o header)
        2. spGetSAFTXMLHeaderDetails - Retorna ProductVersion
        """
        header = SAFTHeader()
        header.FiscalYear = fiscal_year
        header.StartDate = start_date
        header.EndDate = end_date
        header.DateCreated = datetime.now()
        
        conn = PulseConnection.create_connection()
        cursor = conn.cursor()
        
        try:
            # 1. Obter dados principais do spGetSAFTDetails
            print(f"Executando spGetSAFTDetails com LocationCode: {self.location_code}")
            cursor.execute(
                "{CALL dbo.spGetSAFTDetails(?)}",
                self.location_code
            )
            
            # A SP retorna pares Setting/Value
            # Vamos construir um dicionário
            saft_details = {}
            for row in cursor.fetchall():
                setting = row[0]  # Setting
                value = row[1]    # Value
                saft_details[setting] = value
                print(f"  {setting} = {value}")
            
            # Mapear campos do spGetSAFTDetails para o header
            header.AuditFileVersion = saft_details.get('AuditFileVersion', '1.04_01')
            header.CompanyID = saft_details.get('CompanyID', '')
            header.TaxRegistrationNumber = saft_details.get('TaxRegistrationNumber', '')
            header.TaxAccountingBasis = saft_details.get('TaxAccountingBasis', 'F')
            header.CompanyName = saft_details.get('CompanyName', '')
            header.CurrencyCode = saft_details.get('CurrencyCode', 'EUR')
            header.TaxEntity = saft_details.get('TaxEntity', 'Global')
            header.ProductCompanyTaxID = saft_details.get('ProductCompanyTaxID', '')
            header.SoftwareCertificateNumber = saft_details.get('SoftwareCertificateNumber', '')
            header.ProductID = saft_details.get('ProductID', 'PULSE DOMINOS')
            
            # Campos opcionais
            header.HeaderComment = saft_details.get('HeaderComment')
            header.Telephone = saft_details.get('Telephone')
            header.Fax = saft_details.get('Fax')
            header.Email = saft_details.get('Email')
            header.Website = saft_details.get('Website')
            
            # Endereço da empresa
            address = CompanyAddress()
            address.BuildingNumber = saft_details.get('BuildingNumber')
            address.StreetName = saft_details.get('StreetName')
            address.AddressDetail = saft_details.get('AddressDetail', '')
            address.City = saft_details.get('City', '')
            address.PostalCode = saft_details.get('PostalCode', '')
            address.Region = saft_details.get('Region')
            address.Country = saft_details.get('Country', 'PT')
            header.BusinessAddress = address
            
            # 2. Obter ProductVersion do spGetSAFTXMLHeaderDetails
            print(f"\nExecutando spGetSAFTXMLHeaderDetails com LocationCode: {self.location_code}")
            cursor.execute(
                "{CALL dbo.spGetSAFTXMLHeaderDetails(?)}",
                self.location_code
            )
            
            row = cursor.fetchone()
            if row:
                location_code_ret = row[0]  # LocationCode
                version = row[1]            # Version
                header.ProductVersion = version
                print(f"  LocationCode: {location_code_ret}")
                print(f"  Version (ProductVersion): {version}")
            else:
                print("  AVISO: spGetSAFTXMLHeaderDetails não retornou dados")
                header.ProductVersion = "1.0"
            
            print(f"\nHeader montado com sucesso!")
            print(f"  CompanyName: {header.CompanyName}")
            print(f"  ProductID: {header.ProductID}")
            print(f"  ProductVersion: {header.ProductVersion}")
            
        except pyodbc.Error as e:
            print(f"\nERRO SQL ao obter header: {e}")
            raise Exception(f"Erro ao obter header SAFT: {e}")
        finally:
            cursor.close()
            conn.close()
        
        return header
    
    def get_customers(self) -> List[Customer]:
        """Obtém clientes do spGetCustomersForSAFTXML"""
        customers = []
        
        conn = PulseConnection.create_connection()
        cursor = conn.cursor()
        
        try:
            print(f"\nExecutando spGetCustomersForSAFTXML...")
            cursor.execute(
                "{CALL dbo.spGetCustomersForSAFTXML(?)}",
                self.location_code
            )
            
            for row in cursor.fetchall():
                customer = Customer(
                    CustomerID=row[0] or "",
                    AccountID=row[1] or "",
                    CustomerTaxID=row[2] or "",
                    CompanyName=row[3] or "",
                    Contact=row[4] if row[4] else None,
                    Telephone=row[5] if row[5] else None,
                    Fax=row[6] if row[6] else None,
                    Email=row[7] if row[7] else None,
                    Website=row[8] if row[8] else None,
                    SelfBillingIndicator=row[9] if row[9] is not None else 0,
                    BuildingNumber=row[10] if row[10] else None,
                    StreetName=row[11] if row[11] else None,
                    AddressDetail=row[12] or "",
                    City=row[13] or "",
                    PostalCode=row[14] or "",
                    Region=row[15] if row[15] else None,
                    Country=row[16] or "PT"
                )
                customers.append(customer)
            
            print(f"  {len(customers)} clientes extraídos")
        
        except pyodbc.Error as e:
            print(f"ERRO ao obter clientes: {e}")
            raise
        finally:
            cursor.close()
            conn.close()
        
        return customers
    
    def get_products(self) -> List[Product]:
        """Obtém produtos do spGetProductsForSAFTXML"""
        products = []
        
        conn = PulseConnection.create_connection()
        cursor = conn.cursor()
        
        try:
            print(f"\nExecutando spGetProductsForSAFTXML...")
            cursor.execute(
                "{CALL dbo.spGetProductsForSAFTXML(?)}",
                self.location_code
            )
            
            for row in cursor.fetchall():
                product = Product(
                    ProductType=row[0] or "P",
                    ProductCode=row[1] or "",
                    ProductGroup=row[2] if row[2] else None,
                    ProductDescription=row[3] or "",
                    ProductNumberCode=row[1] or ""  # Usar ProductCode como NumberCode
                )
                products.append(product)
            
            print(f"  {len(products)} produtos extraídos")
        
        except pyodbc.Error as e:
            print(f"ERRO ao obter produtos: {e}")
            raise
        finally:
            cursor.close()
            conn.close()
        
        return products
    
    def get_invoices(self, start_date: datetime, end_date: datetime) -> List[Invoice]:
        """Obtém faturas do spGetInvoicesForSAFTXML"""
        invoices = []
        
        conn = PulseConnection.create_connection()
        cursor = conn.cursor()
        
        try:
            print(f"\nExecutando spGetInvoicesForSAFTXML...")
            print(f"  Data início: {start_date.strftime('%Y-%m-%d')}")
            print(f"  Data fim: {end_date.strftime('%Y-%m-%d')}")
            
            cursor.execute(
                "{CALL dbo.spGetInvoicesForSAFTXML(?, ?, ?)}",
                self.location_code,
                start_date.strftime('%Y-%m-%d'),
                end_date.strftime('%Y-%m-%d')
            )
            
            for row in cursor.fetchall():
                invoice = Invoice(
                    InvoiceNo=row[0] or "",
                    ATCUD=row[1] or "0",
                    DocumentStatus=row[2] or "N",
                    Hash=row[3] or "",
                    HashControl=row[4] or "1",
                    Period=row[5] if row[5] is not None else None,
                    InvoiceDate=row[6] if row[6] else datetime.now(),
                    InvoiceType=row[7] or "FT",
                    SourceID=row[8] or "",
                    EACCode=row[9] if row[9] else None,
                    SystemEntryDate=row[10] if row[10] else datetime.now(),
                    CustomerID=row[11] or "",
                    SourceBilling=row[12] if row[12] else None,
                    TaxPayable=Decimal(str(row[13])) if row[13] is not None else Decimal("0"),
                    NetTotal=Decimal(str(row[14])) if row[14] is not None else Decimal("0"),
                    GrossTotal=Decimal(str(row[15])) if row[15] is not None else Decimal("0")
                )
                
                # Carregar linhas da fatura (não implementado aqui - adicione se necessário)
                # invoice.Lines = self.get_invoice_lines(invoice.InvoiceNo)
                
                invoices.append(invoice)
            
            print(f"  {len(invoices)} faturas extraídas")
        
        except pyodbc.Error as e:
            print(f"ERRO ao obter faturas: {e}")
            raise
        finally:
            cursor.close()
            conn.close()
        
        return invoices
